File: lib/py/fenwick_tree.py
# Fenwick Tree (Binary Indexed Tree)
# 1-indexed
class FenwickTree:
    def __init__(self, n):
        self.n = n
        self.tree = [0] * (n + 1)
        
    def update(self, idx, val):
        while idx <= self.n:
            self.tree[idx] += val
            idx += (idx & -idx)
            
    def query(self, idx):
        sum_val = 0
        while idx > 0:
            sum_val += self.tree[idx]
            idx -= (idx & -idx)
        return sum_val
        
    def query_range(self, left, right):
        return self.query(right) - self.query(left - 1)

File: lib/cpp/fenwick_tree.cpp
#include <vector>

using namespace std;

// Fenwick Tree (Binary Indexed Tree)
// 1-indexed 구현
struct FenwickTree {
  int n;
  vector<long long> tree;

  FenwickTree(int n) : n(n) { tree.resize(n + 1); }

  // idx에 val 더하기 (point update)
  void update(int idx, long long val) {
    while (idx <= n) {
      tree[idx] += val;
      idx += (idx & -idx);
    }
  }

  // 1 ~ idx 구간 합 (prefix sum)
  long long query(int idx) {
    long long sum = 0;
    while (idx > 0) {
      sum += tree[idx];
      idx -= (idx & -idx);
    }
    return sum;
  }

  // left ~ right 구간 합
  long long query(int left, int right) {
    return query(right) - query(left - 1);
  }
};
